From 32f3873104faa4323d7db85262145b7895824e4a Mon Sep 17 00:00:00 2001
From: Takayuki 'January June' Suwa <jjsuwa_sys3175@yahoo.co.jp>
Date: Tue, 10 Jan 2023 01:44:09 +0900
Subject: [PATCH] xtensa: Make instruction cost estimation for size more
 accurate

Until now, we applied COSTS_N_INSNS() (multiplying by 4) after dividing
the instruction length by 3, so we couldn't express the difference less
than modulo 3 in insn cost for size (e.g. 11 Bytes and 12 bytes cost the
same).

This patch fixes that.

;; 2 bytes
addi.n	a2, a2, -1	; cost 3

;; 3 bytes
addmi	a2, a2, 1024	; cost 4

;; 4 bytes
movi.n	a3, 80		; cost 5
bnez.n	a2, a3, .L4

;; 5 bytes
srli	a2, a3, 1	; cost 7
add.n	a2, a2, a2

;; 6 bytes
ssai	8		; cost 8
src	a4, a2, a3

:: 3 + 4 bytes
l32r	a2, .L5		; cost 9

;; 11 bytes		; cost 15
;; 12 bytes		; cost 16

gcc/ChangeLog:

	* config/xtensa/xtensa.c (xtensa_insn_cost):
	Let insn cost for size be obtained by applying COSTS_N_INSNS()
	to instruction length and then dividing by 3.
---
 gcc/config/xtensa/xtensa.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/gcc/config/xtensa/xtensa.c b/gcc/config/xtensa/xtensa.c
index b05ae9045..e0adf069e 100644
--- a/gcc/config/xtensa/xtensa.c
+++ b/gcc/config/xtensa/xtensa.c
@@ -4554,13 +4554,15 @@ xtensa_insn_cost (rtx_insn *insn, bool speed)
 {
   if (!(recog_memoized (insn) < 0))
     {
-      int len = get_attr_length (insn), n = (len + 2) / 3;
+      int len = get_attr_length (insn);
 
       if (len == 0)
 	return COSTS_N_INSNS (0);
 
       if (speed)  /* For speed cost.  */
 	{
+	  int n = (len + 2) / 3;
+
 	  /* "L32R" may be particular slow (implementation-dependent).  */
 	  if (xtensa_is_insn_L32R_p (insn))
 	    return COSTS_N_INSNS (1 + xtensa_extra_l32r_costs);
@@ -4607,10 +4609,11 @@ xtensa_insn_cost (rtx_insn *insn, bool speed)
 	    {
 	      /* "L32R" itself plus constant in litpool.  */
 	      if (xtensa_is_insn_L32R_p (insn))
-		return COSTS_N_INSNS (2) + 1;
+		len = 3 + 4;
 
-	      /* Consider ".n" short instructions.  */
-	      return COSTS_N_INSNS (n) - (n * 3 - len);
+	      /* Consider fractional instruction length (for example, ".n"
+		 short instructions or "L32R" litpool constants.  */
+	      return (COSTS_N_INSNS (len) + 1) / 3;
 	    }
 	}
     }
-- 
2.30.2

